dockerGrader/testServer.py

```diff
@@ -0,0 +1,121 @@
+s3 = session.client('s3')
+
+app = Flask(__name__)
+
+logging.basicConfig(
+    handlers=[
+        logging.FileHandler("testServer.log"),
+        logging.StreamHandler()
+    ],
+    level=logging.WARNING)
+
+def downloadSubmission(projectPath):
+    # a project path will look something like this:
+    # projects/p0/users/115799594197844895033/curr.json
+
+    userId = projectPath.split('/')[-2]
+
+    # create user dir for download
+    logging.info('download to {}'.format(CODE_DIR))
+    if os.path.exists(CODE_DIR):
+        shutil.rmtree(CODE_DIR)
+    os.mkdir(CODE_DIR)
+
+    # download
+    response = s3.get_object(Bucket=BUCKET, Key=projectPath)
+    submission = json.loads(response['Body'].read().decode('utf-8'))
+    fileContents = base64.b64decode(submission.pop('payload'))
+    with open(os.path.join(CODE_DIR, submission['filename']), 'wb') as f:
+        f.write(fileContents)
+    with open(os.path.join(CODE_DIR, 'meta.json'), 'w') as f:
+        f.write(json.dumps(submission, indent=2))
+
+def uploadResult(project, studentID):
+    if not os.path.exists("{}/result.json".format(TEST_DIR)):
+        return False
+    with open("result.json", "r") as fr:
+        serializedResult = fr.read()
+    s3.put_object(
+        Bucket=BUCKET,
+        Key='ta/grading/{}/{}.txt'.format(project, studentID),
+        Body=serializedResult.encode('utf-8'),
+        ContentType='text/plain')
+    return True
+
+def lookupGoogleId(netId):
+    path = 'users/net_id_to_google/%s.txt' % netId
+    try:
+        response = s3.get_object(Bucket=BUCKET, Key=path)
+        net_id = response['Body'].read().decode('utf-8')
+        return net_id
+    except botocore.exceptions.ClientError as e:
+        if not e.response['Error']['Code'] == "NoSuchKey":
+            # unexpected error
+            logging.warning(
+                'Unexpected error when look up Googlg ID:' + e.response['Error']['Code'])
+        raise e
+
+def fetchFromS3(project, netId):
+    googleId = lookupGoogleId(netId)
+    if not googleId:
+        return None
+    curPath = ACCESS_PATH.format(project=project, googleId=googleId)
+    downloadSubmission(curPath)
+
+def sendToDocker(project, netId):
+    # create directory to mount inside a docker container
+    if os.path.exists(TEST_DIR):
+        shutil.rmtree(TEST_DIR)
+    shutil.copytree(CODE_DIR, TEST_DIR)
+    shutil.copytree("./io", TEST_DIR + "/io")
+    # we can't use shutil.copytree here again because TEST_DIR exists
+    testCodePath = "./test/{}".format(project)
+    for item in os.listdir(testCodePath):
+        src = os.path.join(testCodePath, item)
+        dest = os.path.join(TEST_DIR, item)
+        if os.path.isdir(src):
+            shutil.copytree(src, dest)
+        else:
+            shutil.copy2(src, dest)
+
+    # run tests inside a docker container
+    image = 'python:3.7-stretch' # TODO: find/build some anaconda image
+
+    cmd = ['docker', 'run',                           # start a container
```

Comment:
I recommend handling timeout with a combination of something like the following:
1. run the container with the -d option (for detached mode).  This will run the container in the background, and output below will be set to the container ID
2. use the "docker wait" command to wait for the container to finish
3. in a separate Python thread, wait three seconds, then issue a "docker rm" to force kill the container and delete it

Reply:
Sounds good. I will update later.